tla-pp/src/main/scala/at/forsyte/apalache/tla/pp/temporal/TableauEncoder.scala

```diff
@@ -0,0 +1,424 @@
+    val newModule = new TlaModule(curModWithPreds.module.name, newDecls)
+
+    curModWithPreds.setModule(newModule)
+  }
+
+  /**
+   * Takes a temporal operator and a name, and generates a new variable declaration declaring the auxiliary "unrolling"
+   * variable for that temporal operator.
+   */
+  private def createUnrollingVar(oper: TlaTempOper, nodeIdentifier: String): TlaVarDecl = {
+    val nameSuffix = oper match {
+      case TlaTempOper.box     => TableauEncoder.BOX_SUFFIX
+      case TlaTempOper.diamond => TableauEncoder.DIAMOND_SUFFIX
+    }
+
+    new TlaVarDecl(nodeIdentifier + nameSuffix)(Typed(BoolT1))
+  }
+
+  /**
+   * Moves down the syntax tree of a provided expression curNode. Each node of the syntax tree that has temporal level,
+   * e.g. contains temporal operators somewhere below it, is encoded by a variable curNode_predicate. For example, for
+   * the formula [](A => <>B), The syntax tree has the shape
+   * {{{
+   *   box
+   *    |
+   *    =>
+   *   /  \
+   *  A   diamond
+   *         |
+   *         B
+   * }}}
+   * Assuming A and B do not contain temporal operators, new variables are introduced for all nodes above them, that is
+   * var_BoxAImpliesDiamondB, var_AImpliesDiamondB, var_DiamondB The value of each variable in a given state corresponds
+   * to a commitment whether or not the formula corresponding to this variable holds true at that point in the trace.
+   * For example, if var_DiamondB is true in a state, the spec will ensure that in some future state, B holds (recall
+   * that B holding at some point in the future is the definition of <>B).
+   */
+  def encodeSyntaxTreeInPredicates(
+      modWithPreds: ModWithPreds,
+      curNode: TlaEx): (ModWithPreds, TlaEx) = {
+    levelFinder.getLevelOfExpression(Set.empty, curNode) match {
+      case TlaLevelTemporal =>
+        curNode match {
+          case NullEx =>
+            throw new IrrecoverablePreprocessingError(
+                "Found a null expression of temporal level; this should not be possible")
+          case NameEx(_) =>
+            throw new IrrecoverablePreprocessingError(
+                "Found a name expression of temporal level.  After inlining no such name expressions should be left in the predicate!")
+          case ValEx(_) =>
+            throw new IrrecoverablePreprocessingError(
+                "Found a value expression of temporal level. This should not be possible.")
+          case LetInEx(_, _) =>
+            throw new IrrecoverablePreprocessingError(
+                "Expect to find no LET-IN expressions. They should have been rewritten by the inliner")
+          case OperEx(oper, args @ _*) =>
+            var curModWithPreds = modWithPreds
+
+            val nodeIdentifier = TableauEncoder.NAME_PREFIX + gen.newName()
+            varNamesToExStrings.addOne(nodeIdentifier, curNode.toString().replace("\"", "\'"))
```

Comment:
Where do the `\"`s come from?

Reply:
Also, see earlier comment on making `varNamesToExStrings` immutable.

Reply:
There could be TlaStrs in the expression. Escaping the quotes avoids terminating strings prematurely, e.g. if this is used in a json. Alternatively, could replace them with escaped quotes (`\\\"`); not sure what to prefer here